
package pborm;


enum OrmMsgOP {
    ORM_COMMAND= 0;
    ORM_INSERT = 1;
    ORM_DELETE = 2;
    ORM_SELECT = 3;
    ORM_UPDATE = 4;
    ORM_COUNT  = 5;
}

message OrmMsgSelectHead {
    optional    int32       offset = 2;
    optional    int32       limit = 3;
}

message OrmMsgReqSelect {
	optional	OrmMsgSelectHead		head = 1;
    optional    string					where = 2;
    optional    string					fields = 4;
    optional    int32					order = 5; //0:nothing todo , -1: asc, 1:desc
	optional	string					orderby = 6;
}

message OrmMsgRspSelect {
	optional	OrmMsgSelectHead		head = 1;
	repeated	bytes					msgs = 2;
}

message OrmMsgReq {
    optional    bytes				msg_data = 1;
    optional    OrmMsgReqSelect     select = 2; //select
}
message OrmMsgRsp {
	optional	int32				err_no = 1; //errno
	optional	string				err_msg = 2; //error msg
	optional	int32				count = 3;//count common
	optional	OrmMsgRspSelect		select = 4;//select
}

message OrmMsgCB {
    optional    bytes          data = 1;
	optional	int64		   cookie = 2;
	optional	uint32		   seq = 3;
}
message OrmMsgOpt {
	optional	uint32		   flatmode = 1[default=0];
}
message OrmMsg {
    required    string          msg_full_type_name = 1;
    required    OrmMsgOP        op  = 2;
	optional	OrmMsgReq		req = 3;
	optional	OrmMsgRsp		rsp = 4;
	optional	OrmMsgCB		cb = 5;
	optional	OrmMsgOpt		opt = 6;
}


////////////////////////////////////////////////////////////////////
